mmgen/models/ddpm/denoising.py

```diff
@@ -0,0 +1,400 @@
+                 dropout=0,
+                 embedding_channels=-1,
+                 num_classes=0,
+                 channels_cfg=None,
+                 output_cfg=dict(mean='eps', var='learned_range'),
+                 norm_cfg=dict(type='GN', num_groups=32),
+                 act_cfg=dict(type='SiLU', inplace=False),
+                 shortcut_kernel_size=1,
+                 use_scale_shift_norm=False,
+                 num_heads=4,
+                 resblock_cfg=dict(type='DenoisingResBlock'),
+                 attention_cfg=dict(type='MultiHeadAttention'),
+                 downsample_conv=True,
+                 upsample_conv=True,
+                 downsample_cfg=dict(type='DenoisingDownsample'),
+                 upsample_cfg=dict(type='DenoisingUpsample'),
+                 attention_res=[16, 8],
+                 pretrained=None):
+
+        super().__init__()
+
+        self.num_classes = num_classes
+        self.num_timesteps = num_timesteps
+        self.rescale_timesteps = rescale_timesteps
+
+        self.output_cfg = deepcopy(output_cfg)
+        self.mean_cfg = self.output_cfg.get('mean', 'eps')
+        self.var_cfg = self.output_cfg.get('var', 'learned_range')
+
+        # double output_channels to output mean and var at same time
+        out_channels = in_channels if self.var_cfg is None else 2 * in_channels
+        self.out_channels = out_channels
+
+        # check type of image_size
+        if not isinstance(image_size, int) and not isinstance(
+                image_size, list):
+            raise TypeError(
+                'Only support `int` and `list[int]` for `image_size`.')
+        if isinstance(image_size, list):
+            assert len(
+                image_size) == 2, 'The length of `image_size` should be 2.'
+            assert image_size[0] == image_size[
+                1], 'Width and height of the image should be same.'
+            image_size = image_size[0]
+        self.image_size = image_size
+
+        channels_cfg_ = deepcopy(self._default_channels_cfg)
+        if channels_cfg is not None:
+            channels_cfg_ = channels_cfg_.update(channels_cfg)
+        if isinstance(channels_cfg_, dict):
+            if image_size not in channels_cfg_:
+                raise KeyError(f'`image_size={image_size} is not found in '
+                               '`channel_cfg`, only support configs for '
+                               f'{[chn for chn in channels_cfg_.keys()]}')
+            self.channel_factor_list = channels_cfg_[image_size]
+        elif isinstance(channels_cfg_, list):
+            self.channel_factor_list = channels_cfg_
+        else:
+            raise ValueError('Only support list or dict for `channel_cfg`, '
+                             f'receive {type(channels_cfg_)}')
+
+        embedding_channels = base_channels * 4 \
+            if embedding_channels == -1 else embedding_channels
+        self.time_embedding = build_module(
+            dict(type='TimeEmbedding'), {
```

Comment:
If you perform a hard-code design, you may directly define an instance of the `TimeEmbedding` instead of using `build_module`.